sklearn/utils/validation.py

```diff
@@ -184,6 +184,43 @@ def _is_arraylike(x):
             hasattr(x, '__array__'))
 
 
+def _num_features(X):
+    """Return the number of features in an array-like X.
+
+    Parameters
+    ----------
+    X : array-like
+        array-like to get the number of features.
+
+    Returns
+    -------
+    features : int
+        Number of features
+    """
+    message = "Unable to find the number of features from X"
+    if not hasattr(X, '__len__') and not hasattr(X, 'shape'):
+        if not hasattr(X, '__array__'):
+            raise TypeError(message)
+        X = np.asarray(X)
+
+    if hasattr(X, 'shape') and X.shape is not None:
+        if len(X.shape) <= 1:
+            raise TypeError(message)
+        if isinstance(X.shape[1], numbers.Integral):
+            return X.shape[1]
```

Comment:
Shall we raise `ValueError` in the `else` branch of this condition? I am not even sure how we could meaningfully trigger this in tests or in a legit usage scenario.

Maybe we should always return `X.shape[1]` without the `if isinstance(X.shape[1], numbers.Integral)` condition.

Reply:
Just for dask dataframes, should we get the first row with the following?

```python
    if hasattr(X, 'iloc'):
        first_elem = X.iloc[0, :]
    else:
        first_elem = X[0]
```

Reply:
I don't think that's necessary: I see no reason why `X.shape[1]` would be delayed in a dask-dataframe.